{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9488b800-22ed-430a-843c-bcdb15e50f78",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from collections import Counter\n",
    "\n",
    "import pandas as pd\n",
    "from thefuzz import fuzz\n",
    "from openpyxl import load_workbook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "82693fcd-97e9-48f9-b009-3ce1fb8c9719",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dataset = \"LUAD\"\n",
    "file_id = \"0052ae83-7ae5-470a-a125-5cd94a9fa9e9\"\n",
    "file_name = \"a6a6b9c6-9db7-42b3-a09f-770b7e126fbb.rna_seq.augmented_star_gene_counts.tsv\"\n",
    "sample_path = f\"./TCGA_GDC/{dataset}/samples_info/{file_id}/{file_name}\"\n",
    "\n",
    "df = pd.read_csv(sample_path, sep=\"\\t\", skiprows=1)\n",
    "df = df.iloc[4:, :]\n",
    "\n",
    "df = df[~df[\"gene_id\"].str.contains(\"_PAR_Y\")]\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "\n",
    "tmp_dict = {gene_name: ensg_id for gene_name, ensg_id in zip(df[\"gene_name\"], df[\"gene_id\"])}\n",
    "gene_ensg_id_dict = {k: tmp_dict[k].split(\".\")[0] for k in sorted(tmp_dict.keys())}\n",
    "\n",
    "json_str = json.dumps(gene_ensg_id_dict, ensure_ascii=False, indent=2)\n",
    "with open(\"./data/gene_ensg_id_dict.json\", \"w\") as f:\n",
    "    f.write(json_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bb9c07bb-fe0f-4a26-bd99-e48a137441ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene list length: 20908\n",
      "[('Y_RNA', 3), ('SFTA3', 2), ('POLR2J3', 2), ('SMIM40', 2), ('PDE11A', 2), ('ACTL10', 2), ('MATR3', 2), ('KCNK3', 1), ('KRTAP5-9', 1), ('SHC4', 1), ('NELL2', 1), ('ACACB', 1), ('RFWD3', 1), ('CALML5', 1), ('PCDH7', 1)]\n",
      "Gene set length: 20900\n"
     ]
    }
   ],
   "source": [
    "with open(\"./data/genes.2023-12-31.json\", \"r\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "\n",
    "print(f\"Gene list length: {len(data)}\")\n",
    "genes = [dic[\"symbol\"] for dic in data]\n",
    "print(Counter(genes).most_common()[:15])\n",
    "\n",
    "genes = {dic[\"symbol\"] for dic in data}\n",
    "print(f\"Gene set length: {len(genes)}\")\n",
    "genes = [gene for gene in genes]\n",
    "genes.sort()\n",
    "\n",
    "with open(\"./data/query_genes.txt\", \"w\") as f:\n",
    "    f.write(\"\\n\".join(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "303a8859-42cd-4bbf-8fdf-3896f25eba94",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"./data/TCGA_Genes.txt\", \"r\") as f:\n",
    "    tcga_genes = {gene for gene in f.read().split(\"\\n\")}\n",
    "\n",
    "with open(\"./data/genes.2023-12-31.json\", \"r\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "gene_to_ensg_id = {dic[\"symbol\"]: dic[\"gene_id\"] for dic in data if dic[\"symbol\"] in tcga_genes}\n",
    "\n",
    "tmp_dict = {\"SFTA3\": \"ENSG00000229415\", \"POLR2J3\": \"ENSG00000168255\", \"SMIM40\": \"ENSG00000286920\",\n",
    "            \"PDE11A\": \"ENSG00000128655\", \"ACTL10\": \"ENSG00000288649\", \"MATR3\": \"ENSG00000015479\"\n",
    "           }\n",
    "gene_to_ensg_id.update(tmp_dict)\n",
    "\n",
    "with open(\"./data/TCGA_Genes.txt\", \"r\") as f:\n",
    "    tcga_genes = f.read().split(\"\\n\")\n",
    "save_gene_to_ensg_id = {gene: gene_to_ensg_id[gene] for gene in tcga_genes}\n",
    "\n",
    "json_str = json.dumps(save_gene_to_ensg_id, ensure_ascii=False, indent=2)\n",
    "with open(\"./data/gene_ensg_id_dict.json\", \"w\") as f:\n",
    "    f.write(json_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7b524fca-3d2f-40e0-b5d7-7a7409c17004",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene ID: 43, Summary Info: Acetylcholinesterase hydrolyzes the neurotransmitter, acetylcholine at neuromuscular junctions and brain cholinergic synapses, and thus terminates signal transmission. It is also found on the red blood cell membranes, where it constitutes the Yt blood group antigen. Acetylcholinesterase exists in multiple molecular forms which possess similar catalytic properties, but differ in their oligomeric assembly and mode of cell attachment to the cell surface. It is encoded by the single ACHE gene, and the structural diversity in the gene products arises from alternative mRNA splicing, and post-translational associations of catalytic and structural subunits. The major form of acetylcholinesterase found in brain, muscle and other tissues is the hydrophilic species, which forms disulfide-linked oligomers with collagenous, or lipid-containing structural subunits. The other, alternatively spliced form, expressed primarily in the erythroid tissues, differs at the C-terminal end, and contains a cleavable hydrophobic peptide with a GPI-anchor site. It associates with the membranes through the phosphoinositide (PI) moieties added post-translationally. AChE activity may constitute a sensitive biomarker of RBC ageing in vivo, and thus, may be of aid in understanding the effects of transfusion[provided by RefSeq, Sep 2019]\n",
      "Gene ID: 2978, Summary Info: This gene encodes an enzyme that plays a role in the recovery of retinal photoreceptors from photobleaching. This enzyme promotes the activity of retinal guanylyl cyclase-1 (GC1) at low calcium concentrations and inhibits GC1 at high calcium concentrations. Mutations in this gene can cause cone dystrophy 3 and code-rod dystrophy 14. provided by RefSeq, Jul 2020]\n",
      "Gene ID: 64135, Summary Info: IFIH1 encodes MDA5 which is an intracellular sensor of viral RNA that triggers the innate immune response. Sensing RNA length and secondary structure, MDA5 binds dsRNA oligonucleotides with a modified DExD/H-box helicase core and a C-terminal domain, thus leading to a proinflammatory response that includes interferons. It has been shown that Coronaviruses (CoVs) as well as various other virus families, are capable of evading the MDA5-dependent interferon response, thus impeding the activation of the innate immune response to infection. MDA5 has also been shown to play an important role in enhancing natural killer cell function in malaria infection. In addition to its protective role in antiviral responses, MDA5 has been implicated in autoimmune and autoinflammatory diseases such as type 1 diabetes, systemic lupus erythematosus, and Aicardi-Goutieres syndrome[provided by RefSeq, Jul 2020]\n"
     ]
    }
   ],
   "source": [
    "with open(\"./data/genes_info.json\", \"r\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "gene_info_dict = {}\n",
    "count_ = 0\n",
    "for gene_id, info in data.items():\n",
    "    if \".[\" in info[\"summary_info\"] or \". [\" in info[\"summary_info\"] or \" [\" in info[\"summary_info\"]:\n",
    "        continue\n",
    "    # if \". [\" not in info[\"summary_info\"] and \" [\" in info[\"summary_info\"]:\n",
    "    #     count_ += 1\n",
    "    #     print(info[\"summary_info\"])\n",
    "    elif \"[provided\" in info[\"summary_info\"] or \" provided \" in info[\"summary_info\"]:\n",
    "        count_ += 1\n",
    "        print(f\"Gene ID: {gene_id}, Summary Info: {info['summary_info']}\")\n",
    "    elif \"NaN\" == info[\"summary_info\"]:\n",
    "        continue\n",
    "    else:\n",
    "        print(\"还有其他情况吗？\")\n",
    "\n",
    "    if count_ > 10:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2b10d843-da00-4970-b50e-1c3af15536c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This gene encodes an enzyme that plays a role in the recovery of retinal photoreceptors from photobleaching. This enzyme promotes the activity of retinal guanylyl cyclase-1 (GC1) at low calcium concentrations and inhibits GC1 at high calcium concentrations. Mutations in this gene can cause cone dystrophy 3 and code-rod dystrophy 14.\n"
     ]
    }
   ],
   "source": [
    "def check_period(your_str):\n",
    "\n",
    "    if your_str[-1] == \".\":\n",
    "        return your_str\n",
    "    else:\n",
    "        return your_str + \".\"\n",
    "\n",
    "\n",
    "summary = \"\"\"This gene encodes an enzyme that plays a role in the recovery of retinal photoreceptors from photobleaching. This enzyme promotes the activity of retinal guanylyl cyclase-1 (GC1) at low calcium concentrations and inhibits GC1 at high calcium concentrations. Mutations in this gene can cause cone dystrophy 3 and code-rod dystrophy 14. provided by RefSeq, Jul 2020]\"\"\"\n",
    "last_index = summary.rfind(\" provided \")\n",
    "summary = check_period(summary[:last_index])\n",
    "print(summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7f284da-6f95-4f88-947f-31d6abc74797",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n",
      "100\n"
     ]
    }
   ],
   "source": [
    "summary = \"\"\"The protein encoded by this gene belongs to the pancreatic ribonuclease family, a subset of the ribonuclease A superfamily. The protein exhibits antimicrobial activity against pathogenic bacteria [provided by RefSeq, Oct 2014]\"\"\"\n",
    "new_summary = \"\"\"The protein encoded by this gene belongs to the pancreatic ribonuclease family, a subset of the ribonuclease A superfamily. The protein exhibits antimicrobial activity against pathogenic bacteria.\"\"\"\n",
    "print(fuzz.partial_ratio(summary, new_summary))\n",
    "\n",
    "summary = \"\"\"This gene encodes a member of the glycosyltransferase family. The specific function of this protein has not been determined. Alternative splicing results in multiple transcript variants of this gene [provided by RefSeq, May 2013]\"\"\"\n",
    "new_summary = \"\"\"This gene encodes a member of the glycosyltransferase family. The specific function of this protein has not been determined. Alternative splicing results in multiple transcript variants of this gene.\"\"\"\n",
    "print(fuzz.partial_ratio(summary, new_summary))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "88aad718-b1f5-4f75-9feb-3da6c8d014ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------------------------------\n",
      "FuzzyWuzzy：简单易用的字符串模糊匹配工具\n",
      "Gene ID: 53827\n",
      "This gene encodes a member of a family of small membrane proteins that share a 35-amino acid signature sequence domain, beginning with the sequence PFXYD and containing 7 invariant and 6 highly conserved amino acids. The approved human gene nomenclature for the family is FXYD-domain containing ion transport regulator. Mouse FXYD5 has been termed RIC (Related to Ion Channel). FXYD2, also known as the gamma subunit of the Na,K-ATPase, regulates the properties of that enzyme. FXYD1 (phospholemman), FXYD2 (gamma), FXYD3 (MAT-8), FXYD4 (CHIF), and FXYD5 (RIC) have been shown to induce channel activity in experimental expression systems. Transmembrane topology has been established for two family members (FXYD1 and FXYD2), with the N-terminus extracellular and the C-terminus on the cytoplasmic side of the membrane. This gene product, FXYD5, is a glycoprotein that functions in the up-regulation of chemokine production, and it is involved in the reduction of cell adhesion via its ability to down-regulate E-cadherin. It also promotes metastasis, and has been linked to a variety of cancers. Alternative splicing results in multiple transcript variants. [RefSeq curation by Kathleen J. Sweadner, Ph.D., sweadner@helix.mgh.harvard.edu., Sep 2009]\n",
      "This gene encodes a member of a family of small membrane proteins that share a 35-amino acid signature sequence domain, beginning with the sequence PFXYD and containing 7 invariant and 6 highly conserved amino acids. The approved human gene nomenclature for the family is FXYD-domain containing ion transport regulator. Mouse FXYD5 has been termed RIC (Related to Ion Channel). FXYD2, also known as the gamma subunit of the Na,K-ATPase, regulates the properties of that enzyme. FXYD1 (phospholemman), FXYD2 (gamma), FXYD3 (MAT-8), FXYD4 (CHIF), and FXYD5 (RIC) have been shown to induce channel activity in experimental expression systems. Transmembrane topology has been established for two family members (FXYD1 and FXYD2), with the N-terminus extracellular and the C-terminus on the cytoplasmic side of the membrane. This gene product, FXYD5, is a glycoprotein that functions in the up-regulation of chemokine production, and it is involved in the reduction of cell adhesion via its ability to down-regulate E-cadherin. It also promotes metastasis, and has been linked to a variety of cancers. Alternative splicing results in multiple transcript variants.\n",
      "10\n",
      "----------------------------------------------------------------------------------------\n",
      "----------------------------------------------------------------------------------------\n",
      "FuzzyWuzzy：简单易用的字符串模糊匹配工具\n",
      "Gene ID: 53822\n",
      "This reference sequence was derived from multiple replicate ESTs and validated by similar human genomic sequence. This gene encodes a member of a family of small membrane proteins that share a 35-amino acid signature sequence domain, beginning with the sequence PFXYD and containing 7 invariant and 6 highly conserved amino acids. The approved human gene nomenclature for the family is FXYD-domain containing ion transport regulator. Transmembrane topology has been established for two family members (FXYD1 and FXYD2), with the N-terminus extracellular and the C-terminus on the cytoplasmic side of the membrane. FXYD2, also known as the gamma subunit of the Na,K-ATPase, regulates the properties of that enzyme. FXYD1 (phospholemman), FXYD2 (gamma), FXYD3 (MAT-8), FXYD4 (CHIF), and FXYD5 (RIC) have been shown to induce channel activity in experimental expression systems. This gene product, FXYD7, is novel and has not been characterized as a protein. [RefSeq curation by Kathleen J. Sweadner, Ph.D., sweadner@helix.mgh.harvard.edu., Dec 2000]\n",
      "This reference sequence was derived from multiple replicate ESTs and validated by similar human genomic sequence. This gene encodes a member of a family of small membrane proteins that share a 35-amino acid signature sequence domain, beginning with the sequence PFXYD and containing 7 invariant and 6 highly conserved amino acids. The approved human gene nomenclature for the family is FXYD-domain containing ion transport regulator. Transmembrane topology has been established for two family members (FXYD1 and FXYD2), with the N-terminus extracellular and the C-terminus on the cytoplasmic side of the membrane. FXYD2, also known as the gamma subunit of the Na,K-ATPase, regulates the properties of that enzyme. FXYD1 (phospholemman), FXYD2 (gamma), FXYD3 (MAT-8), FXYD4 (CHIF), and FXYD5 (RIC) have been shown to induce channel activity in experimental expression systems. This gene product, FXYD7, is novel and has not been characterized as a protein.\n",
      "10\n",
      "----------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "with open(\"./data/genes_info.json\", \"r\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "def check_period(your_str):\n",
    "    if your_str[-1] == \".\":\n",
    "        return your_str\n",
    "    else:\n",
    "        return your_str + \".\"\n",
    "\n",
    "\n",
    "def check_correctness(gene_id, summary, new_summary):\n",
    "    reduce_words = len(summary.split(\" \")) - len(new_summary.split(\" \"))\n",
    "    if reduce_words > 8:\n",
    "        print(\"-\" * 88)\n",
    "        print(\"FuzzyWuzzy：简单易用的字符串模糊匹配工具\")\n",
    "        print(f\"Gene ID: {gene_id}\")\n",
    "        print(summary)\n",
    "        print(new_summary)\n",
    "        print(reduce_words)\n",
    "        print(\"-\" * 88)\n",
    "\n",
    "    if fuzz.partial_ratio(summary, new_summary) < 99.9:\n",
    "        print(\"-\" * 88)\n",
    "        print(\"FuzzyWuzzy：简单易用的字符串模糊匹配工具\")\n",
    "        print(f\"Gene ID: {gene_id}\")\n",
    "        print(summary)\n",
    "        print(new_summary)\n",
    "        print(reduce_words)\n",
    "        print(\"-\" * 88)\n",
    "\n",
    "\n",
    "gene_info_dict = {}\n",
    "for gene_id, info in data.items():\n",
    "    summary = info[\"summary_info\"]\n",
    "\n",
    "    if \".[\" in summary:\n",
    "        last_index = summary.rfind(\".[\")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        continue\n",
    "    elif \". [\" in summary:\n",
    "        last_index = summary.rfind(\". [\")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        continue\n",
    "    elif \" [\" in summary:\n",
    "        last_index = summary.rfind(\" [\")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        continue\n",
    "    elif \"[provided\" in summary:\n",
    "        last_index = summary.rfind(\"[provided\")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        continue\n",
    "    elif \" provided \" in summary:\n",
    "        last_index = summary.rfind(\" provided \")\n",
    "        new_summary = check_period(summary[:last_index])\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "        continue\n",
    "    elif summary == \"NaN\":\n",
    "        new_summary = summary\n",
    "        check_correctness(gene_id, summary, new_summary)\n",
    "        info[\"summary_info\"] = new_summary\n",
    "        gene_info_dict[gene_id] = info\n",
    "\n",
    "json_str = json.dumps(gene_info_dict, ensure_ascii=False, indent=2)\n",
    "with open(\"genes_info_clean.json\", \"w\") as f:\n",
    "    f.write(json_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ec3d0a66-f4d9-4cb7-b7ff-7c6f117f4589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes</th>\n",
       "      <th>Samples</th>\n",
       "      <th>Mutations</th>\n",
       "      <th>Papers</th>\n",
       "      <th>Hyperlinks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>40329</td>\n",
       "      <td>844</td>\n",
       "      <td>913</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACVR1</td>\n",
       "      <td>14329</td>\n",
       "      <td>124</td>\n",
       "      <td>183</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>3787</td>\n",
       "      <td>58</td>\n",
       "      <td>166</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>4905</td>\n",
       "      <td>332</td>\n",
       "      <td>234</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>63618</td>\n",
       "      <td>1533</td>\n",
       "      <td>1238</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>VHL</td>\n",
       "      <td>38699</td>\n",
       "      <td>2984</td>\n",
       "      <td>903</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>WT1</td>\n",
       "      <td>39330</td>\n",
       "      <td>1493</td>\n",
       "      <td>784</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>XPO1</td>\n",
       "      <td>20593</td>\n",
       "      <td>351</td>\n",
       "      <td>338</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>ZFHX3</td>\n",
       "      <td>14801</td>\n",
       "      <td>725</td>\n",
       "      <td>346</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>24627</td>\n",
       "      <td>340</td>\n",
       "      <td>303</td>\n",
       "      <td>https://cancer.sanger.ac.uk/cosmic/gene/analys...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>272 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Genes  Samples  Mutations  Papers  \\\n",
       "0      ABL1    40329        844     913   \n",
       "1     ACVR1    14329        124     183   \n",
       "2    ACVR1B     3787         58     166   \n",
       "3    ACVR2A     4905        332     234   \n",
       "4      AKT1    63618       1533    1238   \n",
       "..      ...      ...        ...     ...   \n",
       "267     VHL    38699       2984     903   \n",
       "268     WT1    39330       1493     784   \n",
       "269    XPO1    20593        351     338   \n",
       "270   ZFHX3    14801        725     346   \n",
       "271   ZRSR2    24627        340     303   \n",
       "\n",
       "                                            Hyperlinks  \n",
       "0    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "1    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "2    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "3    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "4    https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "..                                                 ...  \n",
       "267  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "268  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "269  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "270  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "271  https://cancer.sanger.ac.uk/cosmic/gene/analys...  \n",
       "\n",
       "[272 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# COSMIC - https://cancer.sanger.ac.uk/cosmic/download\n",
    "# Expert Curation of Genes - https://cancer.sanger.ac.uk/cosmic/curation\n",
    "df = pd.read_excel(\"./data/Expert_Curation_of_Genes.xlsx\")\n",
    "wb = load_workbook(\"./data/Expert_Curation_of_Genes.xlsx\")\n",
    "sheet = wb[\"Sheet1\"]\n",
    "\n",
    "hyperlinks = []\n",
    "\n",
    "for row in range(2, len(df) + 2):\n",
    "\n",
    "    hyperlink = sheet.cell(row, 1).hyperlink.target\n",
    "    hyperlinks.append(hyperlink)\n",
    "\n",
    "df[\"Hyperlinks\"] = hyperlinks\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9ea8dd19-5c80-4597-a4b5-8f6d6002ce97",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_excel(\"Expert_Curation_of_Genes.xlsx\", index=False, encoding=\"utf-8\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
